src/app/public/src/modules/shared/affix-top.directive.ts

```diff
@@ -1,39 +1,100 @@
-import { Directive, ElementRef, HostListener, Renderer } from '@angular/core';
+import {
+  AfterViewInit,
+  Directive,
+  ElementRef,
+  HostListener,
+  Renderer
+} from '@angular/core';
+
+import { SkyAppStyleLoader } from '@blackbaud/skyux-builder/runtime';
 
 import { StacheWindowRef } from './window-ref';
 
 @Directive({
   selector: '[stacheAffixTop]'
 })
-export class StacheAffixTopDirective {
-  private componentOffset: number;
+export class StacheAffixTopDirective implements AfterViewInit {
+  public static readonly AFFIX_CLASS_NAME: string = 'stache-affix-top';
+  public isAffixed = false;
+  public isAffixable = false;
+
+  private offsetTop: number = 0;
+  private element: any;
 
   constructor (
     private renderer: Renderer,
     private elementRef: ElementRef,
-    private windowRef: StacheWindowRef) { }
+    private windowRef: StacheWindowRef,
+    private styleLoader: SkyAppStyleLoader) { }
+
+  public ngAfterViewInit(): void {
+    const nativeElement = this.elementRef.nativeElement;
+
+    if (this.isComponent(nativeElement) && nativeElement.children[0]) {
+      this.element = nativeElement.children[0];
+    } else {
+      this.element = nativeElement;
+    }
+
+    this.styleLoader
+      .loadStyles()
+      .then((result: any) => {
+        this.offsetTop = this.element.offsetTop;
+
+        const document = this.windowRef.nativeWindow.document;
+        const documentHeight = document.body.scrollHeight;
+        const viewportHeight = document.documentElement.clientHeight;
+
+        // Prevent the element from removing the viewport scrollbar.
+        const elementOffsetBottom = this.offsetTop + this.element.offsetHeight;
+        this.isAffixable = (documentHeight - elementOffsetBottom > viewportHeight);
+      });
+  }
 
   @HostListener('window:scroll')
-  public onWindowScroll() {
-    if (!this.componentOffset) {
-      this.componentOffset = this.elementRef.nativeElement.offsetTop;
+  public onWindowScroll(): void {
+    if (!this.isAffixable) {
+      return;
     }
 
-    let windowOffset = this.windowRef.nativeWindow.pageYOffset;
+    const windowIsScrolledBeyondElement =
+      (this.offsetTop <= this.windowRef.nativeWindow.scrollY);
 
-    if (this.componentOffset <= windowOffset) {
-      this.setElementClass('stache-affix-top', true);
+    if (windowIsScrolledBeyondElement) {
+      this.affixToTop();
     } else {
-      this.setElementClass('stache-affix-top', false);
+      this.resetElement();
     }
   }
 
-  // private setElementStyle(property: string, value: string) {
-  //   // TODO: Will have to be updated to the renderer2 when we upgrade to Angular 4.x.
-  //   this.renderer.setElementStyle(this.elementRef.nativeElement, property, value);
-  // }
+  private isComponent(element: any): boolean {
+    let isComponent = false;
+
+    Array.prototype.slice.call(element.attributes)
```

Comment:
es 2015 has a nice method so you don't have to do this anymore. 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/of